File: api/src/bigquerytest/java/org/pmiops/workbench/cdr/ConceptBigQueryServiceTest.java
package org.pmiops.workbench.cdr;

import static com.google.common.truth.Truth.assertThat;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.pmiops.workbench.api.BigQueryBaseTest;
import org.pmiops.workbench.api.BigQueryTestService;
import org.pmiops.workbench.db.model.DbCdrVersion;
import org.pmiops.workbench.db.model.DbConceptSetConceptId;
import org.pmiops.workbench.model.Domain;
import org.pmiops.workbench.testconfig.TestJpaConfig;
import org.pmiops.workbench.testconfig.TestWorkbenchConfig;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Import;

@Import({BigQueryTestService.class, ConceptBigQueryService.class, TestJpaConfig.class})
public class ConceptBigQueryServiceTest extends BigQueryBaseTest {

  @Autowired private TestWorkbenchConfig testWorkbenchConfig;
  @Autowired private ConceptBigQueryService conceptBigQueryService;

  @BeforeEach
  public void setUp() {
    DbCdrVersion cdrVersion = new DbCdrVersion();
    cdrVersion.setBigqueryDataset(testWorkbenchConfig.bigquery.dataSetId);
    cdrVersion.setBigqueryProject(testWorkbenchConfig.bigquery.projectId);
    CdrVersionContext.setCdrVersionNoCheckAuthDomain(cdrVersion);
  }

  @Override
  public List<String> getTableNames() {
    return ImmutableList.of("cb_criteria", "cb_search_all_events");
  }

  @Override
  public String getTestDataDirectory() {
    return CB_DATA;
  }

  @Test
  @Disabled("RW-5707")
  public void testGetConceptCountNoConceptsSaved() {
    DbConceptSetConceptId dbConceptSetConceptId1 =
        DbConceptSetConceptId.builder().addConceptId(6L).addStandard(true).build();
    DbConceptSetConceptId dbConceptSetConceptId2 =
        DbConceptSetConceptId.builder().addConceptId(10L).addStandard(false).build();
    assertThat(
            conceptBigQueryService.getParticipantCountForConcepts(
                Domain.CONDITION, ImmutableSet.of(dbConceptSetConceptId1, dbConceptSetConceptId2)))
        .isEqualTo(1);
  }
}

File: api/src/bigquerytest/java/org/pmiops/workbench/cdr/ConceptBigQueryServiceTest.kt
package org.pmiops.workbench.cdr

import com.google.common.truth.Truth.assertThat

import com.google.common.collect.ImmutableSet
import java.util.Arrays
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext
import org.bitbucket.radistao.test.runner.BeforeAfterSpringTestRunner
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.pmiops.workbench.api.BigQueryBaseTest
import org.pmiops.workbench.api.BigQueryService
import org.pmiops.workbench.api.BigQueryTestService
import org.pmiops.workbench.cdr.dao.ConceptDao
import org.pmiops.workbench.cdr.dao.ConceptService
import org.pmiops.workbench.cdr.model.Concept
import org.pmiops.workbench.config.CdrBigQuerySchemaConfigService
import org.pmiops.workbench.db.model.CdrVersion
import org.pmiops.workbench.test.TestBigQueryCdrSchemaConfig
import org.pmiops.workbench.testconfig.TestJpaConfig
import org.pmiops.workbench.testconfig.TestWorkbenchConfig
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Import

@RunWith(BeforeAfterSpringTestRunner::class)
@Import(BigQueryTestService::class, TestBigQueryCdrSchemaConfig::class, TestJpaConfig::class, CdrBigQuerySchemaConfigService::class)
class ConceptBigQueryServiceTest : BigQueryBaseTest() {

    @Autowired
    private val testWorkbenchConfig: TestWorkbenchConfig? = null

    @PersistenceContext
    private val entityManager: EntityManager? = null

    @Autowired
    private val conceptDao: ConceptDao? = null

    @Autowired
    private val bigQueryService: BigQueryService? = null

    @Autowired
    private val cdrBigQuerySchemaConfigService: CdrBigQuerySchemaConfigService? = null

    private var conceptBigQueryService: ConceptBigQueryService? = null

    override val tableNames: List<String>
        get() = Arrays.asList("condition_occurrence")

    override val testDataDirectory: String
        get() = BigQueryBaseTest.MATERIALIZED_DATA

    @Before
    fun setUp() {
        val cdrVersion = CdrVersion()
        cdrVersion.bigqueryDataset = testWorkbenchConfig!!.bigquery!!.dataSetId
        cdrVersion.bigqueryProject = testWorkbenchConfig.bigquery!!.projectId
        CdrVersionContext.setCdrVersionNoCheckAuthDomain(cdrVersion)

        val conceptService = ConceptService(entityManager, conceptDao)
        conceptBigQueryService = ConceptBigQueryService(bigQueryService, cdrBigQuerySchemaConfigService, conceptService)

        conceptDao!!.deleteAll()
    }

    @Test
    fun testGetConceptCountNoConceptsSaved() {
        assertThat(
                conceptBigQueryService!!.getParticipantCountForConcepts(
                        "condition_occurrence", ImmutableSet.of(1L, 6L, 13L, 192819L)))
                .isEqualTo(0)
    }

    @Test
    fun testGetConceptCountConceptsSaved() {
        saveConcept(1L, "S")
        saveConcept(6L, null)
        saveConcept(13L, null)
        saveConcept(192819L, "C")

        assertThat(
                conceptBigQueryService!!.getParticipantCountForConcepts(
                        "condition_occurrence", ImmutableSet.of(1L, 6L, 13L, 192819L)))
                .isEqualTo(2)
    }

    private fun saveConcept(conceptId: Long, standardConceptValue: String?) {
        val concept = Concept()
        concept.conceptId = conceptId
        concept.standardConcept = standardConceptValue
        concept.conceptCode = "concept$conceptId"
        concept.conceptName = "concept $conceptId"
        concept.vocabularyId = "V"
        concept.domainId = "D"
        conceptDao!!.save(concept)
    }
}
